File: src/tool-schema.ts
import type { AgentCLI, ToolDescriptionVariant } from './runtime.js';

/** OpenAI-compatible function tool definition. */
export interface ToolDefinition {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: {
      type: 'object';
      properties: Record<string, unknown>;
      required: string[];
    };
  };
}

export interface BuildToolDefinitionOptions {
  descriptionVariant?: ToolDescriptionVariant;
}

/**
 * Build a tool definition from an AgentCLI instance.
 * The returned object is compatible with OpenAI, Groq, Anthropic,
 * and any provider that accepts the OpenAI tool-calling format.
 */
export function buildToolDefinition(
  cli: AgentCLI,
  toolName = 'run',
  options: BuildToolDefinitionOptions = {},
): ToolDefinition {
  return {
    type: 'function',
    function: {
      name: toolName,
      description: cli.buildToolDescription(options.descriptionVariant),
      parameters: {
        type: 'object',
        properties: {
          command: {
            type: 'string',
            description: 'The CLI command to execute.',
          },
        },
        required: ['command'],
      },
    },
  };
}

File: python/src/onetool/tool_schema.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, TypedDict

from .runtime import AgentCLI, ToolDescriptionVariant


class ToolParameters(TypedDict):
    type: Literal["object"]
    properties: dict[str, object]
    required: list[str]


class ToolFunction(TypedDict):
    name: str
    description: str
    parameters: ToolParameters


class ToolDefinition(TypedDict):
    type: Literal["function"]
    function: ToolFunction


@dataclass(frozen=True, slots=True)
class BuildToolDefinitionOptions:
    description_variant: ToolDescriptionVariant | None = None


def build_tool_definition(
    cli: AgentCLI,
    tool_name: str = "run",
    options: BuildToolDefinitionOptions | None = None,
) -> ToolDefinition:
    resolved = options or BuildToolDefinitionOptions()
    return {
        "type": "function",
        "function": {
            "name": tool_name,
            "description": cli.build_tool_description(resolved.description_variant or "full-tool-description"),
            "parameters": {
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "The CLI command to execute.",
                    }
                },
                "required": ["command"],
            },
        },
    }
